pennylane/_qubit_device.py

```diff
@@ -219,6 +217,13 @@ def execute(self, circuit, **kwargs):
         # increment counter for number of executions of qubit device
         self._num_executions += 1
 
+        # TODO: Remove try/except and consider merging with previous caching
+        # case when circuit is always QuantumTape
+        try:
+            self._circuit_hash = circuit.graph.hash
+        except AttributeError as e:
+            self._circuit_hash = circuit.hash
+
```

Comment:
@josh146 For some reason constructing the `CircuitGraph` of a `QuantumTape` needs to happen after the operations are applied (hence the change in the order of statements within `QubitDevice.execute`.

This phenomenon might also affect caching, as it can happen that we are trying to construct the circuit graph of a tape for the first time just to obtain the hash of the circuit.

The error seems to be related to jacobian computations `ValueError: cannot reshape array of size 4 into shape (2,)`. See [this log](https://github.com/PennyLaneAI/pennylane/runs/1772886525).

Reply:
🤔  That's super weird. As far as I can tell, the tape/graph should not be changing between the old location and the new?

> This phenomenon might also affect caching, as it can happen that we are trying to construct the circuit graph of a tape for the first time just to obtain the hash of the circuit.

Do you mean the code on line 188? All the caching tests pass, so it would seem that that continues to work fine placed there?

Unless the caching tests aren't working correctly...

Reply:
It turns out that by constructing the circuit graph, we mutated underlying tape data (the wires of the observable) which lead to different behaviour.

The wires of the observable are [updated when constructing the `TapeCircuitGraph`](https://github.com/PennyLaneAI/pennylane/blob/c1bfc27d6e191f034ff79ad01cfad2416198ecdc/pennylane/tape/circuit_graph.py#L57) for a quantum function that returns the state. This is a requirement so that the circuit drawer has the specific wires when drawing the circuit.

The update to the wires in the circuit graph, however, affected the tape too (as observables were passed by reference) and propagated back to the `QubitDevice.access_state` method. This method was called with the wires of the observables and it accepts `wires=None` or the specified wires as an argument. If `wires=None`, we get the pure state, otherwise we get the reduced density matrix based on `wires`. What happened was that we ended up getting the reduced density because `wires=None` was not longer the case instead of getting the pure statevector. This lead to incorrect test results.